API-updates

Atticus29 opened the pull request:
# Description

This PR represents the aggregate of a number of smaller PRs that were each approved by either Ed or Nikita. They are collected herein for a final, more thorough review before being officially released.

# Pull Request Checklist:

# Pre-Approval

- [x] There is a description section in the pull request that details what the proposed changes do. It can be very brief if need be, but it ought to exist.
- [ ] Hotfixes should be branched off of the `master` branch and PR'd using the **merge** option (not squashed) into the `hotfix` branch.
- [x] Features and backlog bugs should be merged into the `Development` branch, **NOT** `master`
- [ ] All new text is preferably internationalized (i.e., no end-user-visible text is hard-coded on the PHP pages), and the [spreadsheet tracking internationalizations](https://docs.google.com/spreadsheets/d/133fps9w2pUCEjUA6IGCcQotk7dn9KvepMXJ2IWUZsE8/edit?usp=sharing) has been updated either with a new row or with checkmarks to existing rows.
- [x] There are no linter errors
- [ ] New features have responsive design (i.e., look aesthetically pleasing both full screen and with small or mobile screens)
- [x] [Symbiota coding standards](https://docs.google.com/document/d/1-FwCZP5Zu4f-bPwsKeVVsZErytALOJyA2szjbfSUjmc/edit?usp=sharing) have been followed
- [ ] If any files have been reformatted (e.g., by an autoformatter), the reformat is its own, separate commit in the PR
- [x] Comment which GitHub issue(s), if any does this PR address
    - [https://github.com/orgs/BioKIC/projects/6/views/10?pane=issue&itemId=89420487](https://github.com/orgs/BioKIC/projects/6/views/10?pane=issue&itemId=89420487)
- [ ] If this PR makes any changes that would require additional configuration of any Symbiota portals outside of the files tracked in this repository, make sure that those changes are detailed in [this document](https://docs.google.com/document/d/1T7xbXEf2bjjm-PMrlXpUBa69aTMAIROPXVqJqa2ow_I/edit?usp=sharing).

# Post-Approval

- [ ] It is the code author's responsibility to merge their own pull request after it has been approved
- [ ] If this PR represents a merge into the `Development` branch, remember to use the **squash & merge** option
- [ ] If this PR represents a merge into the `hotfix` branch, remember to use the **merge** option (i.e., no squash).
- [ ] If this PR represents a merge from the `Development` branch into the master branch, remember to use the **merge** option
- [ ] If this PR represents a merge from the `hotfix` branch into the `master` branch use the **squash & merge** option
  - [ ] a subsequent PR from `master` into `Development` should be made with the **merge** option (i.e., no squash).
  - [ ] **Immediately** delete the `hotfix` branch and create a new `hotfix` branch
  - [ ] increment the Symbiota version number in the symbase.php file and commit to the `hotfix` branch.
- [ ] If the dev team has agreed that this PR represents the last PR going into the `Development` branch before a tagged release (i.e., before an imminent merge into the master branch), make sure to notify the team and [lock the `Development` branch](https://github.com/BioKIC/Symbiota/settings/branches) to prevent accidental merges while QA takes place. Follow the release protocol [here](https://github.com/BioKIC/Symbiota/blob/master/docs/release-protocol.md).
- [ ] Don't forget to delete your feature branch upon merge. Ignore this step as required.

Thanks for contributing and keeping it clean!

egbot committed:
Duplicate Search

- Add endpoint for occurrence duplicate search
- Add helper parsing out last name of the full name of a collector
- Add exsiccati data models in support of duplicate search
- Add image OCR to data models in support of duplicate search
- Get rid of obsolete duplicate controller

egbot committed:
Exsiccati schema modifications

- Adjust column and index names to follow the current naming standard

Atticus29 committed:
begin to implement /api/v2/exsiccata/{identifier}/number endpoint

Atticus29 committed:
continue troubleshooting the endpoint

Atticus29 committed:
continue to improve the API endpoint. Fix identifier typo

Atticus29 committed:
confirm that the call works as expected, repopulate the docs, remove cruft

Atticus29 committed:
Api exsiccata identifier (#2261)

* added the endpoint exsiccata/identifier

* indentation fix

* remove redundant casting

* add quotes to HTML response codes

* fix bug where we accidentally removed the array type cast. Sorry, Nikita! -Mark

* make removal of lasteditedby more efficient

---------

Co-authored-by: Nikita Salikov <nsalikov@asu.edu>

NikitaSalikov reviewed (commented):
@Atticus29 Tested and it works! But why the title is ".../number/{numberIdentifier}" if this is /number endpoint?

Atticus29 commented:
> @Atticus29 Tested and it works! But why the title is ".../number/{numberIdentifier}" if this is /number endpoint?

Good call. Yeah, I don't know where those wires got crossed in my brain. Fixed now. Thanks!

Atticus29 committed:
change order of params in /api/v2/exsiccata/{identifier}/number endpoint

Atticus29 committed:
fix the /api/v2/exsiccata/{identifier} endpoint to use eloquent

Atticus29 commented on api/app/Http/Controllers/ExsiccataController.php:
```diff
@@ -68,4 +69,126 @@ public function showAllExsiccata(Request $request) {
+            return response()->json(['error' => 'Record not found'], 404);
+        }
+
+        unset($record->lasteditedby);
```
Fixed in 8061494f5. That problem was that the QueryBuilder doesn't respect the `hidden` array. But switching to stricter eloquent (h/t chatGTP) solved the problem.

Atticus29 commented on api/app/Http/Controllers/ExsiccataController.php:
```diff
@@ -68,4 +69,126 @@ public function showAllExsiccata(Request $request) {
+     *		 required=false,
+     *		 @OA\Schema(type="integer", default=0)
+     *	 ),
+     *	 @OA\Parameter(
```
Fixed in 60830e7ce.

NikitaSalikov committed:
API - exsiccati/{identifier}/number/{numberIdentifier}/occurrence (#2269)

* created endpoint for occurrences in exsiccata/number

* removed the items array in example

NikitaSalikov reviewed (approved):
After changes LGTM too!

Atticus29 committed:
add new exsiccata/{identifier}/number/{numberIdentifier} endpoint (#2290)

* add new exsiccata/{identifier}/number/{numberIdentifier} endpoint

* respond to code review feedback

Atticus29 committed:
Occurrence controller record id accommodation (#2292)

* add a getOccurrence method

* add url decode

* add annotations to include recordID

* improve description for includeMedia param

* respond to code review feedback

Atticus29 committed:
API-audit: Remove search from endpoints (#2302)

* begin to refactor the search endpoints

* remove cruft

* remove commented-out code

NikitaSalikov committed:
API - rename installation endpoints and add API version variable (#2291)

* renamed installation-ping and installation-touch

* added api version variable to environment

* added the guid text to handshake identifier description

* Installation Controller updates

- Convert the "ping" endpoints used within the handshake code to "status" endpoints
- Convert the "touch" endpoints used within the handshake code to "handshake" endpoints

---------

Co-authored-by: Edward Gilbert <egbot@asu.edu>

Atticus29 committed:
Merge branch 'Development' into API-updates

Atticus29 committed:
remove redundant key (#2329)

Atticus29 committed:
rearrange the routes to avoid route parsing conflicts (#2328)

egbot committed:
Merge pull request #2367 from Symbiota/3.3-api-exsiccati-schema

Exsiccati schema modifications

egbot committed:
Merge branch 'api-occurrence-duplicates' into API-updates

egbot committed:
Merge duplicate harvesting endpoint

- Remove duplicate Exsiccati models
- Misc occurrence duplicate controller errors
- Fix issue with route error
- Merge exsiccati db patch statements

NikitaSalikov committed:
added geographic thesaurus helper class, added multiple input support for locality

Atticus29 committed:
resolve merge conflicts

NikitaSalikov committed:
use iso2 instead of numcodes, added user merge with user input not found in db, moved string to array inside helper function

NikitaSalikov committed:
copy occurrenceManager logic into geoHelper function

NikitaSalikov committed:
added grouping to sql where for country

Atticus29 committed:
Occurrence additions (#2456)

* add join to OccurrenceController showAllOccurrences method to add author to the return data

* remove cruft

* add author to getOccurrence method and by extension showOneOccurrence

* make showOneOccurrenceMedia work again

* update showOneOccurrenceIdentifications

* add docblock for showOccurrenceAnnotations endpoint and improve code as well

* begin to self code-review and troubleshoot

* restore identification

* only return the taxa join when the occurrence data is truly needed

* Add more params to the occurrence query (#2470)

* add more params to the occurrence query

* delete files flagged for deletion

* add some options for recordedBy commented out, with my top choice commented in

* remove files accidentally tracked in this PR

* replace join with leftJoin

* replace join with leftJoin for taxon searches as well

* fix left joins to joins

Atticus29 committed:
First pass at collection creation via the api (#2477)

* first pass at collection creation via the api

* not some schema params as required and enum some that have limited options

* add description of individualUrl and disable timestamps on the PHP side

* add TODO

* add logic to handle different types of authentication

Atticus29 committed:
Merge branch 'Development' into API-updates

NikitaSalikov committed:
Merge branch 'API-updates' into API-add-geothesaurus-to-locality

NikitaSalikov committed:
Merge pull request #2426 from Symbiota/API-add-geothesaurus-to-locality

API - add geothesaurus + multiple input support for locality in /occurrence

Atticus29 committed:
Taxon post (#2552)

* begin to implement taxon POST endpoint

* fix forgotten save

* continue to flesh out the taxonomy create method DocBlock

* squashMe

* get the taxa entry to post successfully but not yet taxstatus or full happy path

* I think it is working? Not sure why parenttid is not required

* get the taxstatus api endpoint to work

* remove modifiedUid

* process cultivarEpithet and tradeName before POSTing to taxa

* run makeHidden locally on the taxonomy GET endpoints

Atticus29 committed:
Merge branch 'Development' into API-updates

Atticus29 committed:
Merge branch 'Development' into API-updates

Atticus29 committed:
Morphology api fixes (#2870)

* add enum to morphology endpoint

* remove crufty parameter from showOneCharacter method

Atticus29 committed:
Taxonomy endpoints fixes (#2871)

Atticus29 committed:
add enum to showAllCollections method (#2872)

Atticus29 committed:
add better error handling for showOneCollection method (#2873)

Atticus29 committed:
Improve error handling for installation/{id} endpoint and ocmment out installation/{id}/occurrence for now (#2883)

* improve error handling for installation/{id} endpoint

* comment out installation/{id}/occurrence and save for next release

Atticus29 committed:
add changes https://github.com/Symbiota/Symbiota/pull/2882/files manually without the autoformatting (#2884)

Atticus29 commented:
@egbot just putting this here as I think of it: do we need a way to track which user is making each API call? In other words, should we log all of these API calls?

egbot committed:
Merge pull request #3099 from Symbiota/3.4_rc

Update API dev branch with extended 3.4 Developments

egbot committed:
Merge exsiccati db patch into 3.4 db patch

Atticus29 commented:
> Make sure NOT to Squash Merge, because this PR includes commits from 3 different developers and we want to ensure changes are attributed to each rather than all merged into a single commit.

Done and done.
